labs/cryptol-overview/lab-xx-overview.md

```diff
@@ -0,0 +1,349 @@
+
+You can review the code for yourself on [Amazon's s2n Github Repository](https://github.com/awslabs/s2n). The code relevant to the specification and evaluation of the HMAC routines can be found in the `tests/saw/` directory.
+
+Further exposition on the development of these integration tests can be found in a three part series on the [Galois Inc. Blog](https://galois.com/blog/):
+ * [Part 1](https://galois.com/blog/2016/09/verifying-s2n-hmac-with-saw/) - **Verifying s2n HMAC with SAW**
+ * [Part 2](https://galois.com/blog/2016/09/specifying-hmac-in-cryptol/) - **Specifying HMAC in Cryptol**
+ * [Part 3](https://galois.com/blog/2016/09/proving-program-equivalence-with-saw/) - **Proving Program Equivalence with SAW**
+
+## Microsoft ElectionGuard
+
+ElectionGuard SDK, including key ceremony, ballot encryption, encrypted ballot tally, and partial decryptions for knowledge proofs of trustees.
+
+## Verifying Properties about Algorithms
+
+Cryptol provides an easy interface for using powerful tools such as SAT solvers for verifying properties about algorithms we care about. Throughout this course we will introduce examples and explain how to take advantage of these tools in your own designs and evaluations. Here is an example packaged with the Cryptol source that demonstrates a simple but important property about an encryption algorithm which only uses the (XOR) operation:
+
+```haskell
+encrypt : {a}(fin a) => [8] -> [a][8] -> [a][8]
+encrypt key plaintext = [pt ^ key | pt <- plaintext ]
+
+decrypt : {a}(fin a) => [8] -> [a][8] -> [a][8]
+decrypt key ciphertext = [ct ^ key | ct <- ciphertext ]
+
+property roundtrip k ip = decrypt k (encrypt k ip) == ip
+```
+
+This file defines an `encrypt` operation, a `decrypt` operation, and a property called `roundtrip` which checks for all keys `k` and all input plaintexts `ip` that `decrypt k (encrypt k ip) == ip` (*i.e.* that these operations are inverse to one another).
+
+We can see the effect of encrypting the particular input `attack at dawn` with the key `0xff`:
+
+```haskell
+Main> encrypt 0xff "attack at dawn"
+[0x9e, 0x8b, 0x8b, 0x9e, 0x9c, 0x94, 0xdf, 0x9e, 0x8b, 0xdf, 0x9b,
+ 0x9e, 0x88, 0x91]
+```
+
+Cryptol interprets the string `"attack at dawn"` as a sequence of bytes suitable for the encrypt operations. We will introduce Cryptol types in this lab and discuss them in detail throughout this course.
+
+Furthermore, we can prove this property holds in the interpreter using the `:prove` command and the currently configured SAT solver (Z3 by default):
```

Comment:
It is more precise to say SMT everywhere rather than SAT.

Reply:
Fixed.